In [12]:
using LinearAlgebra, Interpolations, Plots
using HA_solver

# Using HA_solver for Huggett model
Here is a quick demostration of HA_solver, with only implemented (so far hopefully) model: Hugget heterogenous agents with endowments model.
Unfortunately, tuples are fater than sturctures in Julia. That is why I used tuples for writingdown parameters, thus they need to be write down in the set order (or it will throw error or sth worse). In future we can try to use Parameters.jl ...
For now, ehere is a plain notebook with finding q


In [13]:
P = [0.925 0.075
    0.5 0.5]
β = 0.993 
σ = 1.5
e = [1.,.1]
a_n =100
Δ =0.01
g = 0.03
lb = 0.8 #q lower bond for bisection
ub = 1.15 #q upper bound for bisection 
a_min = -2.0 #TODO I need to define a_min twice  
Params = (β,σ, e, P, a_min, 0.8, 1.15)
NumParams = HA_solver.define_NumParam(a_n, Δ, g, a_min, 0.001)
# define asset grid, it is not evenly spaced as each element is computed as $a_grid[i] = a_grid[i-1]+(1+g)^i*Δ $, 
#last value define first element of the gird to be minimal on RHS of the lowest possible point   


(100, [-1.999, -1.9887000000000001, -1.978091, -1.96716373, -1.9559086419, -1.944315901157, -1.93237537819171, -1.920076639537461, -1.9074089387235849, -1.8943612068852924  …  2.5674937076275306, 2.7147885188563565, 2.866502174422047, 3.0227672396547085, 3.18372025684435, 3.349501864549681, 3.5202569204861716, 3.696134628100757, 3.87728866694378, 4.063877326952094], -2.0)

In [14]:
HA_solver.Find_eq_Hugget(Params,NumParams )

Iteration 2
Iteration 3
Iteration 4
Iteration 5
Iteration 6
Iteration 7
Iteration 8
Iteration 9
Iteration 10
Iteration 11
Iteration 12
Iteration 13
Iteration 14
Iteration 15
Iteration 16
Iteration 17
Iteration 18
Iteration 19
Iteration 20
Iteration 21
Iteration 22
Iteration 23
 equilibrium q is 1.012639546394348
DONE


ok 1.0126 is almost 1.0124 in the paper, lest try with -4.0

In [15]:
a_min = -4.0 #TODO I need to define a_min twice  
Params = (β,σ, e, P, a_min, 0.8, 1.15)
NumParams = HA_solver.define_NumParam(a_n, Δ, g, a_min, 0.001)
HA_solver.Find_eq_Hugget(Params,NumParams )

Iteration 2
Iteration 3
Iteration 4
Iteration 5
Iteration 6
Iteration 7
Iteration 8
Iteration 9
Iteration 10
Iteration 11
Iteration 12
Iteration 13
Iteration 14
Iteration 15
Iteration 16
Iteration 17
Iteration 18
Iteration 19
Iteration 20
Iteration 21
Iteration 22
Iteration 23
Iteration 24
Iteration 25
Iteration 26
 equilibrium q is 0.9974296480417253
DONE


Ok this is also close (0.9962) in the article

In [16]:

a_min = -6.0   
Params = (β,σ, e, P, a_min, 0.8, 1.15)
g=0.04
NumParams = HA_solver.define_NumParam(a_n, Δ, g, a_min, 0.005)
HA_solver.Find_eq_Hugget(Params,NumParams )

Iteration 2
Iteration 3
Iteration 4
Iteration 5
Iteration 6
Iteration 7
Iteration 8
Iteration 9
Iteration 10
Iteration 11
Iteration 12
Iteration 13
Iteration 14
Iteration 15
Iteration 16
Iteration 17
Iteration 18
Iteration 19
Iteration 20
Iteration 21
Iteration 22
Iteration 23
Iteration 24
Iteration 25
Iteration 26
Iteration 27
Iteration 28
Iteration 29
Iteration 30
Iteration 31
 equilibrium q is 0.9948722559027374
DONE


For this I needed to change g to have more points on the positive side, I also add minimaly more space from a_min to the grid, 
but the resul is ok


In [18]:
a_min = -8.0   
Params = (β,σ, e, P, a_min, 0.8, 1.15)
g=0.04
NumParams = HA_solver.define_NumParam(a_n, Δ, g, a_min, 0.005)
HA_solver.Find_eq_Hugget(Params,NumParams )

Iteration 2
Iteration 3
Iteration 4
Iteration 5
Iteration 6
Iteration 7
Iteration 8
Iteration 9
Iteration 10
Iteration 11
Iteration 12
Iteration 13
Iteration 14
Iteration 15
Iteration 16
Iteration 17
Iteration 18
Iteration 19
Iteration 20
Iteration 21
Iteration 22
Iteration 23
Iteration 24
Iteration 25
Iteration 26
Iteration 27
Iteration 28
Iteration 29
Iteration 30
Iteration 31
 equilibrium q is 0.9934888784773648
DONE


Thats exactly the same number